### DISCUSSION TITLE
`audit` mode for Renovate CLI tool

### DISCUSSION THREAD

**Discussion by HonkingGoose** [renovatebot/renovate#30019] [Suggest an Idea]
## Summary

Admins who self-host Renovate might benefit from having a built-in `audit` command for the Renovate CLI program. It would:

- Alert if the version of Renovate is lower or equal to a known vulnerable version. Print warning message, recommend user update soon. (Fetch security advisories from `renovatebot/renovate` and compare versions).
- Print version and release dates for user's version and latest version
- Caution if user is using a _really_ old version, say more than 30 days old, recommend user update.

Let me know what you think. :wink: 

#### Audit unsafe version

```bash
$ renovate audit
$
$ WARNING: your version of Renovate has a known vulnerability!
$ Please update Renovate to at least: firstSafeVersion.
$
$ Renovate audit results:
$ 
$ |                             | Version  | Release date | Vulnerable? |
$ | --------------------------- | -------- | ------------ | ----------- |
$ | Your version                | 37.158.0 | 29 Jan 2024  | Yes         |
$ | First safe version          | 37.199.0 | 19 Feb 2024  | No          |
$ | Latest version              | 37.422.4 | 03 July 2024 | No          |
$
$ Renovate audit fail.
```

#### Audit safe, recent version

```bash
$ renovate audit
$
$ Renovate audit results:
$ 
$ |                 | Version number | Release date | Vulnerable? |
$ | --------------- | -------------- | ------------ | ----------- |
$ | Your version    | 37.421.6       | 01 Jul 2024  | No          |
$ | Latest Renovate | 37.422.4       | 03 July 2024 | No          |
$
$ Renovate audit pass.
```

#### Audit user should update, no danger

```bash
$ renovate audit
$
$ Renovate audit results:
$ 
$ |                 | Version number | Released date | Vulnerable? |
$ | --------------- | -------------- | ------------- | ----------- |
$ | Your Renovate   | 37.200.0       | 19 Feb 2024   | No          |
$ | Latest Renovate | 37.422.4       | 03 July 2024  | No          |
$
$ Note: your version of Renovate is x days old.
$ We recommend updating Renovate regularly.
$
$ Renovate audit pass.
```

**Comment by jamietanna** [comment]
@viceice @secustor I think this looks reasonable, what do you think?

**Reply by viceice** [reply]
I'm not sure 🤷‍♂️

**Reply by secustor** [reply]
A WARN log directly after we print the current version would make more sense IMO.

**Reply by HonkingGoose** [reply]
My assumptions:

- Users who self-host will automate Renovate as much as possible.
- Users only read the logs if needed.
- Users don't automatically search for WARN level log output from Renovate in their scripts.

_Only_ printing a WARN log won't nudge the user to update their Renovate program.

With the `audit` command users can create a script like this:

```
1. Run Renovate in audit mode.

    If audit fails: stop script, and print error message.
    If audit passes: continue.

2. Run Renovate in normal update dependencies mode.
```

The goal here is for the computer to say: hey Renovate is old, please update. If we hide things in the logs, users must build their own utility to read the logs and alert on WARN level findings. With the `audit` command users can rely on the outcome of that step to decide what to do next.

**Reply by viceice** [reply]
warnings are visible on dashboard and PR description
